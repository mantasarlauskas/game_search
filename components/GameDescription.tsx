import React, { useEffect, useState } from 'react';
import styles from 'styles/game-description.module.scss';

function GameDescription({ description, slug, letterCount = 500 }: GameDescriptionProps) {
    const [expanded, setExpanded] = useState(false);
    const isOverLimit = description.length > letterCount;

    useEffect(() => {
        setExpanded(false);
    }, [slug]);

    return (
        <div className={styles.root}>
            <div dangerouslySetInnerHTML={{ __html: isOverLimit && !expanded
                ? `${description.substring(0, letterCount)}...` : description }}
            />
            {isOverLimit && (
                <button
                    className={styles.button}
                    type="button"
                    onClick={() => setExpanded((prevExpanded) => !prevExpanded)}
                >
                    {expanded ? 'Show less' : 'Read more'}
                </button>
            )}
        </div>
    );
}

interface GameDescriptionProps {
    description: string;
    letterCount?: number
    slug: string;
}

export default GameDescription;
