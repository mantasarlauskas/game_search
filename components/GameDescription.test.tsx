import GameDescription from 'components/GameDescription';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

describe('<GameDescription />', () => {
    const props = {
        description: 'description',
        slug: 'slug',
    };

    it('matches snapshot', () => {
        const { container } = render(<GameDescription {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('shows description without button', () => {
        render(<GameDescription {...props} />);
        expect(screen.getByText('description')).toBeInTheDocument();
        expect(screen.queryByRole('button')).not.toBeInTheDocument();
    });

    it('shows cropped description with button button', () => {
        render(<GameDescription {...props} letterCount={4} />);
        expect(screen.getByText('desc...')).toBeInTheDocument();
        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('expands and collapses description on click', () => {
        render(<GameDescription {...props} letterCount={4} />);
        expect(screen.getByText('desc...')).toBeInTheDocument();
        userEvent.click(screen.getByRole('button', { name: 'Read more' }));
        expect(screen.getByText('description')).toBeInTheDocument();
        userEvent.click(screen.getByRole('button', { name: 'Show less' }));
        expect(screen.getByText('desc...')).toBeInTheDocument();
    });
});
