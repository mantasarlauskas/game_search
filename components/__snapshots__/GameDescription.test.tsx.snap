// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameDescription /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <div>
      description
    </div>
  </div>
</div>
`;
